### ISSUE TITLE
multihandshake idea

### ISSUE THREAD

**Issue by diasdavid** [diasdavid/node-ipfs-swarm#9]
Right now, for each type of `connection upgrade`, we perform a multistream handshake in order to be able to support several protocols on top of the same transport. This adds a considerable amount of RTT's  to the connection before it is ready, which might represent a problem for slow links. 

0 handshake can bring considerable amounts performance for streaming scenarios. What if we take the multistream handshake and make a special case for a multihandshake such that we send: `/multihandshake/all/the/protocols/that/will/be/stacked/` as one message, instead of handshaking each at all times. This adds the problem that by narrowing down the number of protocols + versions we are interested in talking (muxer1 vs muxer2 or encryptedChannel1 vs encryptedChannel2), we can miss a successful handshake, but for previous known peers, it should work like a charm.

**Comment by jbenet** [comment]
i don't think you need anything different. multistream is designed to do this, you can send everything straight up and achieve a 0-RTT open, given the protocols are known a-priori (or you guess right)
